Fix: Restarting a terminated thread

deepikabhavnani opened the pull request:
### Description
Restarting a terminated thread is a valid use case and should be taken care of.

Resolves: https://github.com/ARMmbed/mbed-os/issues/6930

### Pull request type

    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Breaking change

CC @0xc0170  @c1728p9 @geky

Commit:
Set _finished to false, when resrarting the terminated thread.

Use case:
    Thread      _ISRthread;
    _ISRthread.start(temp_start);
    _ISRthread.join();
    _ISRthread.start(temp_start);
    _ISRthread.join();

Commit:
Set the _tid to NULL before release sempahore

In case task switch happens and join is called, it may fail as
_tid will not be NULL.

Commit:
Semaphore is release by terminate/thunk

0xc0170 commented:
cc @mapellil (as the issue reporter)

0xc0170 commented:
@deepikabhavnani Any tests to prove this fix?

deepikabhavnani commented:
> @deepikabhavnani Any tests to prove this fix?

👍  Sure will add one.

Commit:
Added test case to verify restart of thread

Restarting a terminated thread is a valid use case and is it support is verified
with newly added test case.

pan- commented:
Which OS allows to use their thread API that way ? 
@c1728p9 You did wrote the code (https://github.com/ARMmbed/mbed-os/commit/11ef1d10b8d1cce9fa6ba82e3c241876cb7f8fbc) that make sure a thread cannot be restarted. Is there any hidden trap we don't see ?

Commit:
_tid might change if thread is restarted

Process ID might change, in case of restart from other thread before join
finished.

deepikabhavnani commented:
Had discussion with @c1728p9, as per him synchronization issues and race condition was the limitation.

Below 2 use cases conflict each other:

Case 1: Multiple threads can wait on common thread and should be able to join irrespective of when the thread terminated. In order to achieve this we signal/post semaphore at the end of each join (wait in join). These extra sempahore signals are for threads who join in late for already terminated thread.

If we try to restart the thread, and wait on it: previous post signal will allow main thread to join without even executing the child thread. 

```
/** Test case to verify multiple threads can join to single thread */
template <int N, void (*F)(counter_t *)>
void test_multi_thread_join()
{
    counter_t counter(0);
    Thread threads[N];

    common_thread.start(wait_loop);
    for (int i = 0; i < N; i++) {
        threads[i].start(callback(F, &counter));
    }
    common_thread.terminate();
    for (int i = 0; i < N; i++) {
        threads[i].join();
    }
    TEST_ASSERT_EQUAL(N, counter);
}
```
Case 2: Restart a thread after terminate/join. Since this conflicts with the case 1 we cannot support this.
```
/** Test case to verify restart of thread which terminated successfully */
template <int N, void (*F)(counter_t *)>
void test_single_thread_multiple_times()
{
    counter_t counter(0);
    Thread thread(osPriorityNormal, THREAD_STACK_SIZE);
    
    for (int i = 0; i < N; i++) {
        thread.start(callback(F, &counter));
        thread.join();
    }
    TEST_ASSERT_EQUAL(N, counter);
}
```
Closing this PR and related issue as well stating it will not be supported.
